web-services/security/src/main/java/datawave/security/login/DatawavePrincipalLoginModule.java

```diff
@@ -328,6 +351,29 @@ public boolean login() throws LoginException {
                     throw new AccountLockedException(message);
                 }
             }
+
+            /**
+             * Check terminal server to verify that it can directly connect.
+             * This requires the positive check that the terminal server has an approved role
+             * AuthorizedServer or AuthorizedQueryServer.  If TerminalServer does not have the
+             * correct role we will fail the login.  Currently this only checks for UserType.SERVER.
+             * However the predicate could be modified to include a check for UserType.USER.
+             * Logic just streams through the list of ProxiedUsers to get the last element (terminal server).
+             * Make sure it's not null and if it is a server then we want it to have a direct role.
+             */
+
+            DatawavePrincipal principal = (DatawavePrincipal) getIdentity();
+            DatawaveUser terminalServer = principal.getProxiedUsers().stream().reduce((prev, next) -> next).orElse(null);
+            //terminalUser should never be null, at a minimum the PrincipalUser should be in the chain
+            if(terminalServer == null || (terminalServer.getUserType()== DatawaveUser.UserType.SERVER && !(terminalServer.getRoles().stream().anyMatch(directRoles::contains))))
+            {
+                loginOk = false; // this is critical as it is what the parent class uses to actually deny login
+                String message = "Login denied for terminal server " + terminalServer.getDn() + " due to missing role. Needs one of: " +directRoles+" but has roles: "+terminalServer.getRoles();
```

Comment:
Similarly surprised that `+directRoles+` wasn't formatted as ` + directRoles + `